src/main/java/org/candlepin/subscriptions/resource/api/v2/SubscriptionTableController.java

```diff
@@ -0,0 +1,322 @@
+      skuCapacity.addSubscriptionsItem(invSub);
+      skuCapacity.setQuantity(skuCapacity.getQuantity() + (int) subscription.getQuantity());
+    }
+  }
+
+  private void calculateNextEvent(SubscriptionCapacityView subscription, SkuCapacity skuCapacity) {
+
+    OffsetDateTime nearestEventDate = skuCapacity.getNextEventDate();
+    OffsetDateTime subEnd = subscription.getEndDate();
+    if (subEnd != null
+        && clock.now().isBefore(subEnd)
+        && (nearestEventDate == null || subEnd.isBefore(nearestEventDate))) {
+      nearestEventDate = subEnd;
+      skuCapacity.setNextEventDate(nearestEventDate);
+      skuCapacity.setNextEventType(SubscriptionEventType.SUBSCRIPTION_END);
+    }
+  }
+
+  private static void sortCapacities(
+      List<SkuCapacity> items, SkuCapacityReportSort sort, SortDirection dir) {
+    items.sort(
+        (left, right) -> {
+          var sortField = Optional.ofNullable(sort).orElse(SkuCapacityReportSort.SKU);
+          int sortDir = 1;
+          if (dir == SortDirection.DESC) {
+            sortDir = -1;
+          }
+          int diff =
+              switch (sortField) {
+                case SKU -> left.getSku().compareTo(right.getSku());
+                case SERVICE_LEVEL -> left.getServiceLevel().compareTo(right.getServiceLevel());
+                case USAGE -> left.getUsage().compareTo(right.getUsage());
+                case QUANTITY -> left.getQuantity().compareTo(right.getQuantity());
+                case NEXT_EVENT_DATE -> left.getNextEventDate().compareTo(right.getNextEventDate());
+                case NEXT_EVENT_TYPE -> left.getNextEventType().compareTo(right.getNextEventType());
+                case PRODUCT_NAME -> left.getProductName().compareTo(right.getProductName());
+              };
+          // If the two items are sorted by some other field than SKU and are equal, then break the
+          // tie by sorting by SKU. No two SKUs in the list are equal.
+          if (diff == 0 && sortField != SkuCapacityReportSort.SKU) {
+            diff = left.getSku().compareTo(right.getSku());
+          }
+
+          return diff * sortDir;
+        });
+  }
+
+  private Predicate<SubscriptionCapacityViewMetric> filterByCategory(SkuCapacity inventory) {
+    return m -> {
+      if (inventory.getCategory() != null) {
+        HardwareMeasurementType type = HardwareMeasurementType.fromString(m.getMeasurementType());
+        return type != null && inventory.getCategory().equals(type.toReportCategory());
+      }
+
+      return true;
+    };
+  }
+
+  private static <T> List<T> paginate(List<T> capacities, Pageable pageable) {
+    if (pageable == null) {
+      return capacities;
+    }
+    int offset = pageable.getPageNumber() * pageable.getPageSize();
+    int lastIndex = Math.min(capacities.size(), offset + pageable.getPageSize());
+    return capacities.subList(offset, lastIndex);
+  }
+
+  private static SkuCapacity initializeSkuCapacity(
```

Comment:
Soft suggestion: what do you think of using mapstruct here? Given the field names are mostly the same, it might reduce some boilerplate.

Reply:
+1 to the MapStruct suggestion